src/main/java/at/bitfire/ical4android/util/AndroidTimeUtils.kt

```diff
@@ -197,49 +209,76 @@ object AndroidTimeUtils {
      *
      * @throws ParseException when the string cannot be parsed
      */
-    fun<T: DateListProperty> androidStringToRecurrenceSet(dbStr: String, allDay: Boolean, exclude: Long? = null, generator: (DateList) -> T): T
-    {
-        // 1. split string into time zone and actual dates
-        var timeZone: TimeZone?
-        val datesStr: String
-
-        val limiter = dbStr.indexOf(RECURRENCE_LIST_TZID_SEPARATOR)
-        if (limiter != -1) {    // TZID given
-            val tzId = dbStr.substring(0, limiter)
-            timeZone = DateUtils.ical4jTimeZone(tzId)
-            if (TimeZones.isUtc(timeZone))
-                timeZone = null
-            datesStr = dbStr.substring(limiter + 1)
-        } else {
-            timeZone = null
-            datesStr = dbStr
-        }
-
-        // 2. process date string and generate list of DATEs or DATE-TIMEs
-        val dateList =
-                if (allDay)
-                    DateList(datesStr, Value.DATE)
-                else
-                    DateList(datesStr, Value.DATE_TIME, timeZone)
-
-        // 3. filter excludes
-        val iter = dateList.iterator()
-        while (iter.hasNext()) {
-            val date = iter.next()
-            if (date.time == exclude)
-                iter.remove()
+    fun <T : DateListProperty> androidStringToRecurrenceSet(
+        dbStr: String,
+        allDay: Boolean,
+        exclude: Long? = null,
+        generator: (DateList) -> T
+    ): T {
+        val lines = dbStr.split("\n")
+
+        // extract time zone IDs from the lines
+        val timezones =
+            lines.mapNotNull { line ->
+                val index = line.indexOf(RECURRENCE_LIST_TZID_SEPARATOR)
+                if (index == -1) return@mapNotNull null
+                val tzId = line.substring(0, index)
+                DateUtils.ical4jTimeZone(tzId)
+            }.toSet()           // convert to set to merge duplicates
+
+        // Update dateList's timezone and store if should be adjusted
+        val mainTimeZone: TimeZone =
+            if (timezones.isNotEmpty())
+                timezones.first()
+            else // If size is greater than 1, set to UTC (null)
+                DateUtils.ical4jTimeZone(TimeZones.UTC_ID)!!
```

Comment:
I think it's true that ical4j's DateList is intended to have only one time zone per DateList. Such a list can be used for an RDATE or EXDATE property, which has only one TZID parameter (and all dates in the list have this time zone).

So if we have multiple timezones to handle, I think we should generate a DateList per time zone and change the return signature so that it can return multiple DateLists.

Because now we have only one DateList that contains all values, converted to the same time zone. This is a valid approach, but I think we should do as little "calculation" as possible (there are always things that can go wrong) and return the time zones "as is" instead.

What do you think?